lib/charms/observability_libs/v1/cert_handler.py

```diff
@@ -299,8 +301,16 @@ def __init__(
                 Must match metadata.yaml.
             cert_subject: Custom subject. Name collisions are under the caller's responsibility.
             sans: DNS names. If none are given, use FQDN.
+            refresh_events: an optional list of bound events which
+                will be observed to replace the current CSR with a new one
+                if there are changes in the CSR's DNS SANs or IP SANs.
+                Then, subsequently, replace its corresponding certificate with a new one.
         """
         super().__init__(charm, key)
+        self._stored.set_default(
+            current_sans_ip=None,
+            current_sans_dns=None,
+        )
```

Comment:
We need this to know what our current SANs that generated the current `_csr` to know if we need to generate a new CSR based on the SANs changes.

Reply:
Peer data would assume the peer relation is defined in the charm using the library? Is there a concern about using storedState?

Reply:
in general the feeling is try to avoid it as much as possible

Reply:
in traefik however we're already doing something similar, consider hashing all the bits you want to check for changes instead of storing them individually, it's more extensible https://github.com/canonical/traefik-k8s-operator/blob/main/src/charm.py#L130

Reply:
@sed-i this is blocking a bunch of other work so we'll go ahead with it. we can always open a ticket to refactor using stored data (or finding a way to compare csrs without storing their precursors) if you think it's worth it.